contracts/token/ERC721/ERC721.sol

```diff
@@ -302,19 +302,21 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata {
      * Emits a {Transfer} event.
      */
     function _burn(uint256 tokenId) internal virtual {
-        address owner = ERC721.ownerOf(tokenId);
+        address from = ERC721.ownerOf(tokenId);
+
+        _beforeTokenTransfer(from, address(0), tokenId);
 
-        _beforeTokenTransfer(owner, address(0), tokenId);
+        require(from == ERC721.ownerOf(tokenId), "ERC721: burn from incorrect owner");
```

Comment:
Instead of this require, why don't we just burn from whoever the owner is after `_beforeTokenTransfer`? After all, the function signature isn't requesting that we burn this from a specific owner.

We should add a comment why we're reading the owner again after the hook.

Reply:
If we do that, and if the `_beforeTokenTransfer` does a transfer, we end up with inconsistent events:

```
Transfer(<original>, <receiver>, tokenId);
Transfer(<original>, address(0), tokenId);
```

When we should have

```
Transfer(<original>, <receiver>, tokenId);
Transfer(<receiver>, address(0), tokenId);
```

Reply:
Also, the `_balances[from] -= 1;` is no longer correct.

Reply:
We could however replace this require with a refresh of the owner.

```
address from = ERC721.ownerOf(tokenId);
```

Reply:
That's what I meant. To refresh the owner.

Reply:
done